### ISSUE TITLE
PostgreSQL Metrics Support

### ISSUE THREAD

**Issue by NoraCodes** [linkerd/linkerd2#2632]
### What problem are you trying to solve?

Deep inspection into PostgreSQL connections and the data being passed across them would provide more granularity in debugging performance issues for many database-driven applications.

### How should the problem be solved?

Implement a PostgreSQL codec that allows the Linkerd proxy to inspect the PostgreSQL wire protocol and provide metrics for the communications with the database server. This could be exposed as a CLI as well as providing Prometheus metrics.

Potentially useful metrics made available as a result of this:

- Open database connections
- Authentication failures/successes (more than a very few failures is indicative of either a very bad misconfiguration, or malice)
- Connection lifetime
- Query-response latency
- Response size

### Related Issues

As #1779, similar to #2214.

**Comment by NoraCodes** [comment]
### Target

Enable inspection of PostgreSQL wire protocol by the Linkerd2-proxy

### Current status

PostgreSQL is treated as an unknown protocol so only very basic TCP metrics are available

### Problems for implementation

1. A method is required for recognizing when a PostgreSQL connection is being made. This method must be relatively free of overhead so as not to increase overall proxy latency.
1. An async-friendly PostgreSQL wire protocol decoder is required.
1. Both methods must be integrated into the proxy in an error-free, low-latency way.

### Design questions

1. How much should Linkerd be responsible for connecting connection-based data to the actual information in the database (e.g. schemata?) (I think probably this is not Linkerd's responsibility, but we could make some steps toward this if we wish.)
1. How do we handle PostgreSSL? (Should we ask users to let us terminate SSL for them? Probably not, but we could.)
1. What additional metrics should we expose for users? (My inclination is to implement what is fast and obvious and let users ask for more.)

### Implementation Plans

1. Prototype, test, and characterize a basic PostgreSQL connection detection mechanism with **a fixed-size buffer per connection** and low latency.
1. Working with other implementing additional proxy protocol handlers, build an extensible and low-latency method of integration of **an arbitrary number of protocol handlers using the same buffer**. Specifically, I envision this capability as a buffer per connection filled from incoming TCP packets, which is the large enough for the largest protocol handler's trigger system. As packets come in, more and more handlers' triggers will be able to count themselves out, until eventually only one or zero will be left, at which point handling will be passed to that handler, or the connection will be considered as an unknown protocol. This is essentially an **online algorithm for protocol detection** (for each protocol x, given these n bytes, is it possible this is the beginning of a valid connection of type x).
1. Characterize and test this algorithm extensively, with a focus on keeping latency down.
1. Integrate this trigger with the HTTP1, HTTP2, and other protocols in the proxy.
1. Using `tokio` (already used by Linkerd2-proxy) and [`nom`](https://github.com/Geal/nom) parser combinators, build a performant and extensible MVP of a PostgreSQL wire protocol parser. This should start as a true MVP which can be easily extended to produce more comprehensive introspection and metrics.
1. Characterize and test each feature implemented in the parser, specifically looking for pathological cases where valid PostgreSQL input leads to an explosion of runtime in the parser, thus producing high latency.
1. Integrate this parser and trigger with the proxy, creating the necessary metrics reporting as needed.

### Existing Work

I plan to base as much of the parser for the PostgreSQL wire protocol as possible on existing code, including [`tokio-postgres`](https://github.com/sfackler/rust-postgres/tree/master/tokio-postgres)

**Comment by grampelberg** [comment]
> How much should Linkerd be responsible for connecting connection-based data to the actual information in the database (e.g. schemata?) (I think probably this is not Linkerd's responsibility, but we could make some steps toward this if we wish.)

Not sure what you're thinking here, have any suggestions?

> How do we handle PostgreSSL? (Should we ask users to let us terminate SSL for them? Probably not, but we could.)

If we don't terminate TLS, the stream will be encrypted and we won't be able to do much with it. Definitely going to need to do the mTLS dance that we're doing with HTTP.

> What additional metrics should we expose for users? (My inclination is to implement what is fast and obvious and let users ask for more.)

Getting to the golden metrics is the MVP I think (success rate, latency, qps). I'm sure there's a ton that we could come up with past that, but getting towards something akin to service profiles (type such as select/insert) would be great eventually.

> Implementation Plans

I'd love to see:

- CLI integration of the stats called out. My hypothesis is that `stat` won't work particularly well and we'll see a different subcommand. Dashboard and grafana can be totally kept out of scope.

@hawkw you know the proxy way better than I do, how does this all sound to you?

**Comment by hawkw** [comment]
>     2. Working with others implementing additional proxy protocol handlers, build an extensible and low-latency method of integrating **an arbitrary number of protocol handlers using the same buffer**. Specifically, I envision this capability as a buffer per connection filled from incoming TCP packets, which is the large enough for the largest protocol handler's trigger system. As packets come in, more and more handlers' triggers will be able to count themselves out, until eventually only one or zero will be left, at which point handling will be passed to that handler, or the connection will be considered as an unknown protocol. This is essentially an **online algorithm for protocol detection** (for each protocol x, given these n bytes, is it possible this is the beginning of a valid connection of type x).

Right now the proxy's protocol detection logic is essentially a form of this algorithm that's hard coded to only detect HTTP/1 and HTTP/2, so this seems like an appropriate approach.

**Comment by NoraCodes** [comment]
> > How much should Linkerd be responsible for connecting connection-based data to the actual information in the database (e.g. schemata?) (I think probably this is not Linkerd's responsibility, but we could make some steps toward this if we wish.)
> 
> Not sure what you're thinking here, have any suggestions?

Well, in theory, we could have a method by which users could register schemata with Linkerd, or we could even intercept it on the fly as it's created, assuming we are intercepting all Postgres connections throughout the lifetime of the database. Then, we could call out statistics by table, by column, etc with pie charts and other fancy things. Definitely out of scope for the MVP, but it might factor into our design process.

> Right now the proxy's protocol detection logic is essentially a form of this algorithm that's hard coded to only detect HTTP/1 and HTTP/2, so this seems like an appropriate approach.

I expect it will be pretty easy to make fast protocol triggers for HTTP/1 and HTTP/2 protocols, too, so that'll be rolled into this.

**Comment by hawkw** [comment]
Something else that comes to mind is whether we might eventually want the proxy to do more than just collect metrics --- i.e., could we manage connection pooling et cetera? Of course we wouldn't want to do that for a MVP but it's an interesting thought.

**Comment by grampelberg** [comment]
> Well, in theory, we could have a method by which users could register schemata with Linkerd, or we could even intercept it on the fly as it's created, assuming we are intercepting all Postgres connections throughout the lifetime of the database. Then, we could call out statistics by table, by column, etc with pie charts and other fancy things. Definitely out of scope for the MVP, but it might factor into our design process.

Yeah! This is exactly what I was thinking about around service profile integration.

> could we manage connection pooling et cetera? 

If we do the service profile/schema stuff, a lot of those patterns come across as well I believe. Retries, timeouts, load balancing between backends, connection pooling and the like.

**Comment by NoraCodes** [comment]
> I'd love to see [...] CLI integration of the stats called out. My hypothesis is that `stat` won't work particularly well and we'll see a different subcommand. Dashboard and grafana can be totally kept out of scope.

I spent a lot of today looking at this and it honestly seems like, from an interface perspective, at least the basic stats can definitely be presented in `stat` as it is. However, as time goes on, it might be worth building a pluggable interface for command line statistics reporting, especially if going forward we want to support Kafka, Redis, other databases, and more exotic things (stream stats being a big one the current `stat` interface doesn't look good for).
